tree/tree/src/TBasket.cxx

```diff
@@ -1035,16 +1146,26 @@ Int_t TBasket::WriteBuffer()
          nzip   += kMAXZIPBUF;
       }
       nout = noutot;
+      fKeylen += compoffkey; // before creating TKey, we need to reserve enough space for compression entry offsets.
       Create(noutot,file);
+      fKeylen -= compoffkey; // reduce fKeylen back to the original length (without compression entry offsets).
       fBufferRef->SetBufferOffset(0);
-
-      Streamer(*fBufferRef);         //write key itself again
-      memcpy(fBuffer,fBufferRef->Buffer(),fKeylen);
+      Bool_t random = fRandomAccessCompression;
+      if (random) {
+         fRandomAccessCompression = 0;// make sure compressed entry offsets are not streamed into uncompressed buffer
+         Streamer(*fBufferRef);         //write key itself again
+         memcpy(fBuffer,fBufferRef->Buffer(),fKeylen);
+         fRandomAccessCompression = 1;// streaming compressed entry offsets into compressed buffer
+         fLast += compoffkey; // fLast is different between fBufferRef and fCompressedBufferRef. It needs to reserve compressed entry offsets for fCompressedBufferRef.
+         Streamer(*fCompressedBufferRef);
+      } else {
+         Streamer(*fBufferRef);         //write key itself again
+         memcpy(fBuffer,fBufferRef->Buffer(),fKeylen);
+      }         
    } else {
       fBuffer = fBufferRef->Buffer();
       Create(fObjlen,file);
       fBufferRef->SetBufferOffset(0);
-
```

Comment:
Spurious changes, please revert - stopping here; please clean up the patch to not blow up the size by unrelated / unmotivated white space changes.